File: app/SourcesShared/DetailMetaRecoveryPolicy.swift
import Foundation

/// Pure decisions used by both detail surfaces while recovering metadata for catalog IDs that the
/// installed meta add-ons cannot answer directly. Kept free of app types so the shipped policy can
/// be compiled into a small standalone regression test.
enum DetailMetaRecoveryPolicy {
    enum Resolution: String, Equatable {
        case ready
        case pending
        case unresolved
    }

    enum EntryState: Equatable {
        case ready
        case loading
        case failed
        case notStarted
    }

    static func resolution(entries: [EntryState]) -> Resolution {
        if entries.contains(.ready) { return .ready }
        if entries.isEmpty { return .unresolved }
        if entries.contains(.loading) || entries.contains(.notStarted) { return .pending }
        return .unresolved
    }

    /// A terminal result is meaningful only for the request currently selected by the engine.
    /// Returning nil for a selection mismatch prevents a stale title's empty result from triggering
    /// recovery on the page that has just replaced it.
    static func resolution(
        selectedID: String?,
        requestedID: String,
        entries: [EntryState]
    ) -> Resolution? {
        guard selectedID == requestedID else { return nil }
        return resolution(entries: entries)
    }

    enum TMDBMedia: String, Equatable {
        case movie
        case tv
    }

    enum CatalogIDShape: Equatable {
        case imdb(String)
        case tmdb(Int, media: TMDBMedia?)
        case tvdb(Int)
        /// A Kitsu anime id (`kitsu:460`). Unlike the others this carries no id TMDB understands, so it
        /// resolves in TWO hops: Kitsu's own mappings API names the title's TheTVDB / TMDB / IMDb id, and
        /// that shape then resolves exactly like a native `tvdb:` / `tmdb:` catalog id. Anime add-ons hand
        /// out these ids constantly, and without a shape for them every anime detail page opened from such a
        /// catalog had no recovery path at all and latched on "Details unavailable".
        case kitsu(Int)
        case unsupported
    }

    /// Parse only catalog ID shapes with an unambiguous lookup path. Unknown schemes remain
    /// unsupported instead of being guessed onto a possibly unrelated title.
    static func catalogIDShape(_ raw: String) -> CatalogIDShape {
        let value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty else { return .unsupported }

        if let imdb = imdbID(value) { return .imdb(imdb) }

        let parts = value.split(separator: ":", omittingEmptySubsequences: false).map(String.init)
        if let embedded = parts.compactMap(imdbID).first { return .imdb(embedded) }

        let lowered = parts.map { $0.lowercased() }
        if lowered.count == 2, lowered[0] == "tmdb" {
            guard let id = positiveInteger(lowered[1]) else { return .unsupported }
            return .tmdb(id, media: nil)
        }
        if lowered.count == 3, lowered[0] == "tmdb" {
            guard let id = positiveInteger(lowered[2]),
                  let kind = TMDBMedia(rawValue: lowered[1]) else {
                return .unsupported
            }
            return .tmdb(id, media: kind)
        }
        if lowered.count == 2, lowered[0] == "tvdb" {
            guard let id = positiveInteger(lowered[1]) else { return .unsupported }
            return .tvdb(id)
        }
        // Strict on purpose, exactly like `tvdb:` above: the SHOW id only. An episode-qualified anime id
        // ("kitsu:460:1:2") is not a detail-page route, and accepting its trailing numbers would be the
        // guessing this parser exists to prevent.
        if lowered.count == 2, lowered[0] == "kitsu" {
            guard let id = positiveInteger(lowered[1]) else { return .unsupported }
            return .kitsu(id)
        }
        if lowered.count == 1 {
            guard let id = positiveInteger(lowered[0]) else { return .unsupported }
            return .tmdb(id, media: nil)
        }
        return .unsupported
    }

    private static func imdbID(_ raw: String) -> String? {
        let lowered = raw.lowercased()
        guard lowered.hasPrefix("tt"), lowered.count > 2 else { return nil }
        let digits = lowered.dropFirst(2)
        guard digits.allSatisfy(asciiDigit) else { return nil }
        return lowered
    }

    private static func positiveInteger(_ raw: String) -> Int? {
        guard !raw.isEmpty, raw.allSatisfy(asciiDigit), let value = Int(raw), value > 0 else {
            return nil
        }
        return value
    }

    private static func asciiDigit(_ character: Character) -> Bool {
        character >= "0" && character <= "9"
    }
}

File: android/app/src/main/kotlin/com/vortx/android/catalog/DetailMetaRecoveryPolicy.kt
package com.vortx.android.catalog

/// Pure decisions used while recovering metadata for catalog ids that the installed meta add-ons cannot
/// answer directly (a `tmdb:`/`tvdb:`/`kitsu:` id whose Cinemeta meta never resolves). The Kotlin port of
/// Apple `app/SourcesShared/DetailMetaRecoveryPolicy.swift`, kept free of app types so it can be reasoned
/// about (and tested) on its own. The detail ViewModel uses [catalogIdShape] to decide an id is a non-`tt`
/// shape worth a one-shot recovery via `TMDBPersonClient.imdbId`, and [Resolution] to describe how the meta
/// load ended.
object DetailMetaRecoveryPolicy {

    /// How a meta request ended for the purposes of the terminal "Details unavailable" decision. Mirrors
    /// Apple `DetailMetaRecoveryPolicy.Resolution`.
    enum class Resolution { READY, PENDING, UNRESOLVED }

    /// The recognised shapes of a detail-page catalog id. Only shapes with an unambiguous lookup path are
    /// parsed; an unknown scheme stays [Unsupported] rather than being guessed onto an unrelated title.
    /// Mirrors Apple `DetailMetaRecoveryPolicy.CatalogIDShape`.
    enum class TmdbMedia { MOVIE, TV }

    sealed interface CatalogIdShape {
        data class Imdb(val id: String) : CatalogIdShape
        data class Tmdb(val id: Int, val media: TmdbMedia?) : CatalogIdShape
        data class Tvdb(val id: Int) : CatalogIdShape
        data class Kitsu(val id: Int) : CatalogIdShape
        data object Unsupported : CatalogIdShape
    }

    /// Parse only catalog id shapes with an unambiguous lookup path (`tt…`, `tmdb:[movie|tv:]N`, `tvdb:N`,
    /// `kitsu:N`, and a bare positive integer treated as a tmdb id). Everything else is [Unsupported].
    /// Mirrors Apple `DetailMetaRecoveryPolicy.catalogIDShape`.
    fun catalogIdShape(raw: String): CatalogIdShape {
        val value = raw.trim()
        if (value.isEmpty()) return CatalogIdShape.Unsupported

        imdbId(value)?.let { return CatalogIdShape.Imdb(it) }

        val parts = value.split(":")
        parts.firstNotNullOfOrNull { imdbId(it) }?.let { return CatalogIdShape.Imdb(it) }

        val lowered = parts.map { it.lowercase() }
        if (lowered.size == 2 && lowered[0] == "tmdb") {
            val id = positiveInteger(lowered[1]) ?: return CatalogIdShape.Unsupported
            return CatalogIdShape.Tmdb(id, media = null)
        }
        if (lowered.size == 3 && lowered[0] == "tmdb") {
            val id = positiveInteger(lowered[2]) ?: return CatalogIdShape.Unsupported
            val kind = tmdbMedia(lowered[1]) ?: return CatalogIdShape.Unsupported
            return CatalogIdShape.Tmdb(id, media = kind)
        }
        if (lowered.size == 2 && lowered[0] == "tvdb") {
            val id = positiveInteger(lowered[1]) ?: return CatalogIdShape.Unsupported
            return CatalogIdShape.Tvdb(id)
        }
        // Strict on purpose, exactly like `tvdb:` above: the SHOW id only. An episode-qualified anime id
        // ("kitsu:460:1:2") is not a detail-page route.
        if (lowered.size == 2 && lowered[0] == "kitsu") {
            val id = positiveInteger(lowered[1]) ?: return CatalogIdShape.Unsupported
            return CatalogIdShape.Kitsu(id)
        }
        if (lowered.size == 1) {
            val id = positiveInteger(lowered[0]) ?: return CatalogIdShape.Unsupported
            return CatalogIdShape.Tmdb(id, media = null)
        }
        return CatalogIdShape.Unsupported
    }

    /// True when [raw] is a non-`tt` id whose shape has an unambiguous recovery lookup path, so the detail
    /// ViewModel should attempt a one-shot `TMDBPersonClient.imdbId` recovery for it rather than sitting on
    /// a permanent skeleton.
    fun isRecoverableNonImdb(raw: String): Boolean = when (catalogIdShape(raw)) {
        is CatalogIdShape.Imdb, CatalogIdShape.Unsupported -> false
        else -> true
    }

    private fun imdbId(raw: String): String? {
        val lowered = raw.lowercase()
        if (!lowered.startsWith("tt") || lowered.length <= 2) return null
        val digits = lowered.drop(2)
        return if (digits.all { it in '0'..'9' }) lowered else null
    }

    private fun tmdbMedia(raw: String): TmdbMedia? = when (raw) {
        "movie" -> TmdbMedia.MOVIE
        "tv" -> TmdbMedia.TV
        else -> null
    }

    private fun positiveInteger(raw: String): Int? {
        if (raw.isEmpty() || !raw.all { it in '0'..'9' }) return null
        val value = raw.toIntOrNull() ?: return null
        return if (value > 0) value else null
    }
}
